src/Agent.Worker/TaskRunner.cs

```diff
@@ -56,10 +56,11 @@ public async Task RunAsync()
             var handlerFactory = HostContext.GetService<IHandlerFactory>();
 
             // Set the task id and display name variable.
-            ExecutionContext.Variables.Set(Constants.Variables.Task.DisplayName, DisplayName);
-            ExecutionContext.Variables.Set(WellKnownDistributedTaskVariables.TaskInstanceId, Task.Id.ToString("D"));
-            ExecutionContext.Variables.Set(WellKnownDistributedTaskVariables.TaskDisplayName, DisplayName);
-            ExecutionContext.Variables.Set(WellKnownDistributedTaskVariables.TaskInstanceName, Task.Name);
+            var scope =  ExecutionContext.Variables.CreateScope();
+            scope.Set(Constants.Variables.Task.DisplayName, DisplayName);
+            scope.Set(WellKnownDistributedTaskVariables.TaskInstanceId, Task.Id.ToString("D"));
+            scope.Set(WellKnownDistributedTaskVariables.TaskDisplayName, DisplayName);
+            scope.Set(WellKnownDistributedTaskVariables.TaskInstanceName, Task.Name);
```

Comment:
no, the scope passes this change to the Variables object that created it, it just remembers what it did so when it is Disposed(), it can undo it. Right now it only supports addition of new variables, but could be updated to remember the previous value and the reset it upon disposal as well

Reply:
maybe my naming could be better, if you have a better suggestion of a name of the object/variable, I'm all ears